--- a/configgen/generators/libretro/libretroConfig.py	2024-05-19 18:51:26.239056252 +0000
+++ b/configgen/generators/libretro/libretroConfig.py	2024-05-19 18:51:35.079030774 +0000
@@ -15,6 +15,7 @@
 import utils.videoMode as videoMode
 import controllersConfig
 import xml.etree.ElementTree as ET
+from pathlib import Path
 
 eslog = get_logger(__name__)
 sys.path.append(
--- a/configgen/generators/libretro/libretroConfig.py	2024-04-12 02:03:40.836960712 +0000
+++ b/configgen/generators/libretro/libretroConfig.py	2024-05-09 19:08:23.036133015 +0000
@@ -562,6 +563,58 @@
             controller, pad = controller_list[i - 1]
             if (pad.guid in valid_n64_controller_guids and pad.configName in valid_n64_controller_names) or (system.isOptSet(f'{option}-controller{i}') and system.config[f'{option}-controller{i}'] != 'retropad'):
                 update_n64_controller_config(i)
+    
+    ## TATE mode remap for handhelds
+    if system.config['core'] in ['fbneo', 'mame']:                     
+        path = Path(rom)
+        folder_name = path.parent.name   
+            
+        handhelds = {
+            ('19000000010000000100000000010000', 'Deeplay-keys'): {  # rg35xx/h
+                'rotation': 'left', 
+                'remap': {
+                    'stk_r_x+': '18', 'stk_r_x-': '19', 'stk_r_y+': '17', 'stk_r_y-': '16',
+                    'btn_a': '0', 'btn_b': '1', 'btn_x': '8', 'btn_y': '-1',
+                }
+            },
+            ('19000000010000000100000000010000', 'odroidgo2 joypad'): {  # rg28xx
+                'rotation': 'left', 
+                'remap': {
+                    'btn_down': '7', 'btn_left': '5', 'btn_right': '4', 'btn_up': '6',
+                    'btn_start': '0', 'btn_select': '8', 'btn_l2': '1', 'btn_r': '2',
+                    'btn_r2': '3', 'btn_a': '-1', 'btn_b': '-1', 'btn_x': '-1', 'btn_y': '-1',
+                }
+            },
+            ('030000005e0400008e02000014010000', 'TRIMUI Player1'): {  # TrimUI
+                'rotation': 'left', 
+                'remap': {
+                    'stk_r_x+': '18', 'stk_r_x-': '19', 'stk_r_y+': '17', 'stk_r_y-': '16',
+                    'btn_a': '0', 'btn_b': '1', 'btn_x': '8', 'btn_y': '-1',
+                }
+            },
+        }
+            
+        def update_handheld_config(guid, name):
+            if (guid, name) in handhelds:
+                settings = handhelds[(guid, name)]
+                # set display rotation
+                if settings['rotation'] == 'left':
+                    if (system.config['core'] == 'fbneo'):
+                        coreSettings.save('fbneo-vertical-mode', '"' + 'TATE alternate' + '"')
+                    elif (system.config['core'] == 'mame'):
+                        coreSettings.save('mame_rotation_mode', '"' + 'tate-rol' + '"')
+                elif settings['rotation'] == 'right':
+                    if (system.config['core'] == 'fbneo'):
+                        coreSettings.save('fbneo-vertical-mode', '"' + 'TATE' + '"')
+                    elif (system.config['core'] == 'mame'):
+                        coreSettings.save('mame_rotation_mode', '"' + 'tate-ror' + '"')
+                # remap inputs
+                for btn, value in settings['remap'].items():
+                    retroarchConfig[f'input_player1_{btn}'] = value    
+               
+        controller, pad = sorted(controllers.items())[0] 
+        if (system.isOptSet(f"{systemCore}-hhtate") and system.config[f"{systemCore}-hhtate"] == "True" or folder_name == "tate"):
+            update_handheld_config(pad.guid, pad.configName)
 
     ## PORTS
     ## Quake
